components/fxa-client/ios/FxAClient/FxAccountDeviceConstellation.swift

```diff
@@ -0,0 +1,192 @@
+            }
+        } catch {
+            FxALog.error("Failure ensuring device capabilities: \(error).")
+        }
+    }
+
+    public func setLocalDeviceName(name: String) {
+        DispatchQueue.global().async {
+            do {
+                try self.handleAccountAuthErrors { acct in
+                    try acct.setDeviceDisplayNameSync(name)
+                }
+                self.refreshState()
+            } catch {
+                FxALog.error("Failure changing the local device name: \(error).")
+            }
+        }
+    }
+
+    public func pollForEvents(completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void) {
+        DispatchQueue.global().async {
+            do {
+                let events = try self.handleAccountAuthErrors { acct in
+                    try acct.pollDeviceCommandsSync()
+                }
+                DispatchQueue.main.async { completionHandler(.success(events)) }
+            } catch {
+                DispatchQueue.main.async { completionHandler(.failure(error)) }
+            }
+        }
+    }
+
+    public func sendEventToDevice(targetDeviceId: String, e: DeviceEventOutgoing) {
+        DispatchQueue.global().async {
+            do {
+                switch e {
+                case let .sendTab(title, url): do {
+                    try self.handleAccountAuthErrors { acct in
+                        try acct.sendSingleTabSync(targetId: targetDeviceId, title: title, url: url)
+                    }
+                }
+                }
+            } catch {
+                FxALog.error("Error sending event to another device: \(error).")
+            }
+        }
+    }
+
+    public func setDevicePushSubscription(sub: DevicePushSubscription) {
+        // No need to wrap in async, this operation doesn't do any IO or heavy processing.
+        do {
+            try handleAccountAuthErrors { acct in
+                try acct.setDevicePushSubscriptionSync(
+                    endpoint: sub.endpoint,
+                    publicKey: sub.publicKey,
+                    authKey: sub.authKey
+                )
+            }
+        } catch {
+            FxALog.error("Failure setting push subscription: \(error).")
+        }
+    }
+
+    public func processRawIncomingDeviceEvent(pushPayload: String,
+                                              completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void) {
+        DispatchQueue.global().async {
+            do {
+                let events = try self.handleAccountAuthErrors { acct in
+                    try acct.handlePushMessageSync(payload: pushPayload)
+                }
+                DispatchQueue.main.async { completionHandler(.success(events)) }
+            } catch {
+                DispatchQueue.main.async { completionHandler(.failure(error)) }
+            }
+        }
+    }
+
+    internal func handleAccountAuthErrors<T>(_ cb: (FirefoxAccount) throws -> T) rethrows -> T {
```

Comment:
401 errors are fed back to the manager (which pumps the state machine/tries to recover in response) through notifications.